File: Source/Binary.FSharp/Creators.Collections/ListConverter.fs
namespace Mikodev.Binary.Creators.Collections

open Mikodev.Binary
open System

type internal ListConverter<'a>(converter : Converter<'a> , memoryConverter : Converter<Memory<'a>>) =
    inherit Converter<List<'a>>(0)

    override __.Encode(allocator, item) =
        if not (obj.ReferenceEquals(item, null)) then
            for i in item do
                converter.EncodeAuto(&allocator, i)
        ()

    override __.Decode(span : inref<ReadOnlySpan<byte>>) : List<'a> =
        let memory = memoryConverter.Decode &span
        let span = memory.Span
        let mutable list = []
        for i = span.Length - 1 downto 0 do
            list <- span.[i] :: list
        list

File: code/Binary/Internal.SpanLike/ListConverter.cs
namespace Mikodev.Binary.Internal.SpanLike;

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

internal sealed class ListConverter<E>(Converter<E> converter) : Converter<List<E>>
{
    private readonly Converter<E> converter = converter;

    public override void Encode(ref Allocator allocator, List<E>? item)
    {
        var converter = this.converter;
        var source = CollectionsMarshal.AsSpan(item);
        foreach (var i in source)
            converter.EncodeAuto(ref allocator, i);
    }

    public override List<E> Decode(in ReadOnlySpan<byte> span)
    {
        return SpanLikeMethods.GetList(this.converter, span);
    }
}
